src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php

```diff
@@ -0,0 +1,465 @@
+use Symfony\Component\Routing\Route;
+
+/**
+ * Helps add and import routes into a RouteCollection.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class RouteCollectionBuilder
+{
+    private $loader;
+
+    /**
+     * A mixture of different objects that hold routes.
+     *
+     * @var Route[]|RouteCollectionBuilder[]
+     */
+    private $routes = array();
+
+    private $defaults = array();
+
+    private $prefix;
+
+    private $host;
+
+    private $condition;
+
+    private $requirements = array();
+
+    private $options = array();
+
+    private $schemes;
+
+    private $methods;
+
+    private $resources = array();
+
+    private $controllerClass;
+
+    /**
+     * @param LoaderInterface $loader
+     */
+    public function __construct(LoaderInterface $loader)
+    {
+        $this->loader = $loader;
+    }
+
+    /**
+     * Import an external routing resource, like a file.
+     *
+     * Returns a RouteCollectionBuilder so you can continue to tweak options on the routes.
+     *
+     * @param mixed  $resource
+     * @param string $prefix
+     * @param string $type
+     *
+     * @return RouteCollectionBuilder
+     */
+    public function import($resource, $prefix = null, $type = null)
+    {
+        /** @var RouteCollection $subCollection */
+        $subCollection = $this->resolve($resource, $type)->load($resource, $type);
+
+        return $this->addRouteCollection($subCollection, $prefix);
+    }
+
+    /**
+     * Adds a route and returns it for future modification.
+     *
+     * @param string      $path       The route path
+     * @param string      $controller The route controller string
+     * @param string|null $name       The name to give this route
+     *
+     * @return Route
+     */
+    public function add($path, $controller, $name = null)
+    {
+        $route = new Route($path);
+        $route->setDefault('_controller', $controller);
+        $this->addRoute($route, $name);
+
+        return $route;
+    }
+
+    /**
+     * Returns a RouteCollectionBuilder that can be configured and then added with addBuilder().
+     *
+     * @param string $prefix A prefix to apply to all routes added to this collection
+     *
+     * @return RouteCollectionBuilder
+     */
+    public function createBuilder($prefix = null)
+    {
+        $builder = new self($this->loader);
+        $builder->setPrefix($prefix);
+
+        return $builder;
+    }
+
+    /**
+     * Add a RouteCollectionBuilder.
+     *
+     * @param RouteCollectionBuilder $builder
+     */
+    public function addBuilder(RouteCollectionBuilder $builder)
+    {
+        $this->routes[] = $builder;
+    }
+
+    /**
+     * Adds a RouteCollection directly and returns those routes in a RouteCollectionBuilder.
+     *
+     * @param RouteCollection $collection
+     * @param string|null     $prefix
+     *
+     * @return $this
+     */
+    public function addRouteCollection(RouteCollection $collection, $prefix = null)
```

Comment:
I would remove this one from the public API; the `createBuilder()` one is probably enough. If not, why would I use a builder instead of creating a route collection directly?